File: frontend/src/app/account.service.ts
import { HttpClient, HttpParams } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { Observable } from "rxjs";
import { Account } from "./account";
import { environment } from "../environments/environment";

@Injectable({ providedIn: 'root' })
export class AccountService {
  private apiServerUrl = environment.apiBaseUrl;

  constructor(private http: HttpClient) {}

  // Get the account with the given account number and pin
  // Used for login
  public getAccountByAccountNumberAndPin(account: Account): Observable<Account> {
    let params = new HttpParams();
    params = params.set('accountNumber', account.accountNumber);
    params = params.set('pinNumber', account.pinNumber) ;
    return this.http.get<Account>(`${this.apiServerUrl}/account/get-account`, { params });
  }

  // Create the given account
  public addAccount(account: Account): Observable<Account> {
    return this.http.post<Account>(`${this.apiServerUrl}/account/add`, account);
  }

  // Update the given account
  public updateAccount(account: Account): Observable<Account> {
    return this.http.put<Account>(`${this.apiServerUrl}/account/update`, account);
  }

  // Delete the given account
  // Not currently used
  public deleteAccount(accountNumber: string): Observable<void> {
    return this.http.delete<void>(`${this.apiServerUrl}/account/delete/${accountNumber}`);
  }
}

File: backend/src/main/java/com/example/atmproject/service/AccountService.java
package com.example.atmproject.service;

import java.util.Optional;

import javax.security.auth.login.AccountNotFoundException;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.example.atmproject.model.Account;
import com.example.atmproject.repo.AccountRepo;

@Service
public class AccountService {
  private final AccountRepo accountRepo;

  @Autowired
  public AccountService(AccountRepo accountRepo) {
    this.accountRepo = accountRepo;
  }

  // Get the account with the given account number and pin if one exists
  public Optional<Account> getAccountByAccountNumberAndPin(String accountNumber, String pinNumber) throws AccountNotFoundException {
    return accountRepo.findAccountByAccountNumberAndPinNumber(accountNumber, pinNumber);
  }

  // Create the given account if it doesn't already exist
  public Account addAccount(Account account) {
    if (!accountRepo.findAccountByAccountNumber(account.getAccountNumber()).isEmpty()) return null;
    else return accountRepo.save(account);
  }

  // Update the given account
  public Account updateAccount(Account account) {
    return accountRepo.save(account);
  }

  // Delete the given account
  public void deleteAccount(String accountNumber) {
    accountRepo.deleteAccountByAccountNumber(accountNumber);
  }
}
